### ISSUE TITLE
Node port parameter is not suppoorted using kubernetes plugin

### ISSUE THREAD

**Issue by cmoulliard** [hashicorp/waypoint#718]
## Issue

The following hcl config which includes the `node_port` parameter to configure the kubernetes k8s resource as mentioned within the doc - https://www.waypointproject.io/plugins/kubernetes  is not supported

**hcl config**
```
project = "example-java"

app "example-java" {
    build {
        use "pack" {
            builder="gcr.io/buildpacks/builder:v1"
        }
        registry {
              use "docker" {
                image = "localhost:5000/example-java"
                tag   = "latest"
                local = true
              }
            }
    }

    deploy {
      use "kubernetes" {
        namespace = "demo"
        probe_path = "/"
        service_port = 8080
        node_port = 32000
      }
    }

    release {
      use "kubernetes" {
      }
    }
}
```

**Error**
```
waypoint deploy 

» Deploying...
! /Users/cmoullia/temp/waypoint/waypoint-examples/docker/java/waypoint.hcl:22,9-18:
  Unsupported argument; An argument named "node_port" is not expected here.

```

**Comment by crisp2u** [comment]
I believe `node_port ` should go under release stanza

**Comment by dominikbraun** [comment]
Right, it works for me when `service_port` is in the `deploy` stanza and `node_port` in the `release` stanza.

**Comment by dominikbraun** [comment]
Just checked out the documentation - `node_port` is documented under `kubernetes (releasemanager)` correctly: https://www.waypointproject.io/plugins/kubernetes#node_port

**Comment by cmoulliard** [comment]
Why do we have to define the `node_port` parameter under `release` ? Sorry but I'm really confused about the purpose of what you call a `release` ? For me, when I deploy a kubernetes application, I expect that the service and/or ingress route will be deployed without the need to configure something part of `release` @dominikbraun

**Comment by nicholasjackson** [comment]
@cmoulliard waypoint separates the concept of deploy and release. Deploy is the state when your app is in the environment and running but not accepting traffic. Released is when it is actively accepting traffic. With this approach, you could do something like deploying an application to prod and run some final acceptance checks on it to validate it is ok before adding it to the production traffic.

**Comment by cmoulliard** [comment]
I disagree as we need to test an application on kubernetes to have the service exposed like the ingress route, otherwise it will be impossible for a developer using kind, k3s or minikube to test/debug the microservice deployed

**Comment by mitchellh** [comment]
What Nic said is right, but also what you said makes sense: there needs to be some alternate means to view a deployment and a node port is an option. I think right now we’ve been leaning heavily on our URL service to provide this for us but having valid options that don’t require that are sensible.

**Comment by cmoulliard** [comment]
The `release` concept (which is a concept that pivotal/red hat/ibm/...) dont support (yet) is interesting for different reasons as it could be used to perform additional steps as by example :
 
- Package the microservice as helm chart for deployments on test/prod clusters
- Build the image with a certified OCI image (including CVEs, ...) as replacement of the test/debug image used by thedeveloper during build/deploy test
- Generate RBAC files to re-enforce the control over the serviceaccounts/roles able to play with the application
- Populate the final URL if the application is controlled by a loadbalancer ...

**Comment by mitchellh** [comment]
Indeed, conceptually it can do many things. What we're focused on in the short term is enabling blue/green-type patterns, integrating with service meshes, etc. The initial idea is: open traffic to the end user.

**Comment by cmoulliard** [comment]
`servicemesh` is too complex for the developers as no tool exists to generate the istio kubernetes manifests. The only project available is this one - https://github.com/snowdrop/istio-java-api but not integrated part of Dekorate - https://dekorate.io to generate the manifest files.

So, ideally the `publish` step should be then focused to deploy a kubernetes application (deployment, service, serviceaccount, secret, confimap) and make the URL available using an ingress enpoint and `release` step be focused around more ops tasks

**Comment by mitchellh** [comment]
> servicemesh is too complex for the developers as no tool exists to generate the istio kubernetes manifests.

Exactly, that's true across non-kubernetes use cases generally, too. I've described service mesh in its current form as a "syscall" layer and we're missing the more usable userspace. That was actually _the_ motivation initially behind the `release` concept separation in Waypoint. :)

**Comment by cmoulliard** [comment]
Do you agree for the moment to move back the `node_part` parameter under the `deploy` section to let then the waypoint server to create a service using or not a node_port and to allow a developer to consume the service while it uses `deploy` ?

Ideally an ingress route should be created too for k8s users and openshift route for openshift users
